source/contributed/typescript-and-screeps-2.md

```diff
@@ -0,0 +1,598 @@
+
+```typescript
+// worker.ts
+export abstract class Worker {
+  public creep: Creep;
+  public memory: WorkerMemory;  // use the new 'WorkerMemory'
+}
+
+// miner.ts
+export class Miner extends Worker {
+  public memory: MinerMemory;  // use the new 'MinerMemory'
+}
+```
+
+OK, at first this looks like it will work — but actually you will get errors.  When you redeclare `memory` on the child class (`Miner`), the parent no longer has access to it, so any methods in `Worker` that call `this.memory` will throw an error or get `undefined`.
+
+We can fix by telling the parent about the child-class's extended Memory using generics:
+
+```typescript
+// worker.ts
+// this next line does 3 things:
+// 1. we declare a generic type M (short for "memory"),
+// 2. we tell the compiler that M is any type extending from WorkerMemory
+// 3. we set a default value for M as WorkerMemory
+export abstract class Worker<M extends WorkerMemory = WorkerMemory> {
+  public creep: Creep;
+  // now we tell TypeScript that `memory` property is of type `M`.
+  // that is, `memory` is any type extending from `WorkerMemory`,
+  // and defaulting to `WorkerMemory` if not specified
+  public memory: M;
+}
+
+// miner.ts
+// if you think of `M` above as a variable (it is), here we have to assign
+// a value to `M`; so we tell TypeScript, "for the `Miner` child class, `M` is
+// type `MinerMemory`" (since we set a default above, if we leave this out
+// TypeScript will just treat `M` as type `WorkerMemory`)
+export class Miner extends Worker<MinerMemory> {
+  // we no longer need to redefine memory here, so we  can remove this line
+  // public memory: MinerMemory;  // no longer needed
+}
+```
+
+Using generics like this well help Types just "flow" through the system as you code, and the compiler will be able to infer everything it needs.
+
+### Generic Custom Screeps Objects & Type Guards
+
+We covered how to use extending interfaces and generics to keep your Memory objects organized.  Now lets look at how to use them to keep the objects in the Screeps World organized.  We wan't to gather as much "type" information about objects as we can, so we can leverage the advantages of the TypeScript typechecker when writing code.
+
+As you know, the objects in Memory are reset each tick.  If you try and store an object in Memory, next tick it is invalid.  So we have to store the object ID, and next tick:
+```typescript
+// `myObject` will be type of {}
+const myObject = Game.getObjectById(objectID);
+```
+However, when we get the object back, we lost it's type.  It could be anything.  In the case of `Miner` above, *we know* it's a Source — but the compiler doesn't.
+
+We can tell it, though.  The `getObjectById` method is actually typed as a generic, so we can use:
+```typescript
+// `myObject` will be type of Source
+const myObject = Game.getObjectById<Source>(objectID);
+```
+
+This works good enough for that specific example.  But we want MOAR.
```

Comment:
I don't think we actually want "MOAR". As funny as it sounds, in my opinion the article should be informative.